tests/unit/h/views/api/groups_test.py

```diff
@@ -8,477 +8,606 @@
 
-@pytest.mark.usefixtures(
-    "CreateGroupAPISchema", "group_service", "group_create_service"
-)
-class TestCreateGroup:
-    def test_it_inits_group_create_schema(self, pyramid_request, CreateGroupAPISchema):
-        views.create(pyramid_request)
+        with pytest.raises(ValidationError):
+            views.create(pyramid_request)
 
-        CreateGroupAPISchema.return_value.validate.assert_called_once_with({})
-
-    # @TODO Move this test once _json_payload() has been moved to a reusable util module
-    def test_it_raises_if_json_parsing_fails(self, pyramid_request):
-        """It raises PayloadError if parsing of the request body fails."""
-        # Make accessing the request.json_body property raise ValueError.
-        type(pyramid_request).json_body = {}
-        with mock.patch.object(
-            type(pyramid_request), "json_body", new_callable=mock.PropertyMock
-        ) as json_body:
-            json_body.side_effect = ValueError()
-            with pytest.raises(views.PayloadError):
-                views.create(pyramid_request)
-
-    def test_it_passes_request_params_to_group_create_service(
-        self, pyramid_request, CreateGroupAPISchema, group_create_service
+    @pytest.mark.parametrize(
+        "appstruct,group_create_service_method_call",
+        [
+            (
+                {"name": sentinel.name, "description": sentinel.description},
+                call.create_private_group(
+                    name=sentinel.name,
+                    userid=sentinel.userid,
+                    description=sentinel.description,
+                    groupid=None,
+                ),
+            ),
+            (
+                {"name": sentinel.name},
+                call.create_private_group(
+                    name=sentinel.name,
+                    userid=sentinel.userid,
+                    description=None,
+                    groupid=None,
+                ),
+            ),
+        ],
+    )
+    def test_create_private_group(
+        self,
+        pyramid_request,
+        CreateGroupAPISchema,
+        group_create_service,
+        appstruct,
+        group_create_service_method_call,
+        assert_it_returns_group_as_json,
     ):
-        CreateGroupAPISchema.return_value.validate.return_value = {
-            "name": "My Group",
-            "description": "How about that?",
-        }
-        views.create(pyramid_request)
+        CreateGroupAPISchema.return_value.validate.return_value = appstruct
+
+        response = views.create(pyramid_request)
 
-        group_create_service.create_private_group.assert_called_once_with(
-            "My Group",
-            pyramid_request.user.userid,
-            description="How about that?",
-            groupid=None,
+        assert group_create_service.method_calls == [group_create_service_method_call]
+        assert_it_returns_group_as_json(
+            response,
+            group=group_create_service.create_private_group.return_value,
```

Comment:
All the views also contain this repetitive code to pass a group through `GroupJSONPresenter` and return the resulting dict:

```python
    return GroupJSONPresenter(group, request).asdict(expand=["organization", "scopes"])
```

Again repetitive view code results in repetitive tests as for each view we need to test that `GroupJSONPresenter` was initialized with the right arguments, `asdict()` was called with the right arguments, and the result of `asdict()` was returned.

This time I went with a test helper method as otherwise the repetition in the test code is really massive.

Again, the way this should work is just by specifying a [renderer](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/renderers.html) in the view config:

```python
@view_config(..., renderer="group_json")
```

Then the view would just return the group object directly and Pyramid would handle passing it through the renderer, and the view's tests wouldn't need to bother with asserting about rendering.